File: scripts/flutter/openapi/update_pubspec.py
#!/usr/bin/env python3
"""패키지 pubspec.yaml에 BuiltValue 의존성을 자동 추가하는 스크립트.

built_value, built_collection (dependencies)과
build_runner, built_value_generator (dev_dependencies)를 추가합니다.
멱등성 보장: 이미 있으면 스킵합니다.
"""

import re
import sys
from pathlib import Path

# ──────────────────────────────────────────────
# Constants
# ──────────────────────────────────────────────

DEPENDENCIES = {
    "built_value": "^8.12.3",
    "built_collection": "^5.1.1",
}

DEV_DEPENDENCIES = {
    "build_runner": "^2.4.15",
    "built_value_generator": "^8.12.3",
}


# ──────────────────────────────────────────────
# YAML text manipulation helpers
# ──────────────────────────────────────────────


def _find_block_last_entry(content: str, block_name: str) -> int:
    """지정된 블록의 마지막 항목 인덱스를 반환합니다."""
    lines = content.split("\n")
    in_block = False
    last_entry_idx = -1

    for i, line in enumerate(lines):
        if re.match(rf"^{block_name}:", line):
            in_block = True
            continue
        if in_block:
            if line.startswith("  "):
                last_entry_idx = i
            elif line.strip() and not line.startswith(" "):
                break

    return last_entry_idx


def _has_dependency(content: str, dep_name: str) -> bool:
    """의존성이 이미 존재하는지 확인합니다."""
    return bool(re.search(rf"^\s+{re.escape(dep_name)}:", content, re.MULTILINE))


def _add_dependency(content: str, block_name: str, dep_name: str, version: str) -> str:
    """의존성을 블록에 추가합니다."""
    last_idx = _find_block_last_entry(content, block_name)
    if last_idx == -1:
        return content

    lines = content.split("\n")
    lines.insert(last_idx + 1, f"  {dep_name}: {version}")
    return "\n".join(lines)


# ──────────────────────────────────────────────
# Public API
# ──────────────────────────────────────────────


def update_pubspec(pubspec_path: Path, dry_run: bool = False) -> bool:
    """pubspec.yaml에 BuiltValue 관련 의존성을 추가합니다.

    Args:
        pubspec_path: pubspec.yaml 파일 경로.
        dry_run: True이면 미리보기만.

    Returns:
        True if any change was made.
    """
    if not pubspec_path.exists():
        print(f"Error: {pubspec_path} not found", file=sys.stderr)
        return False

    content = pubspec_path.read_text(encoding="utf-8")
    original = content
    added: list[str] = []

    # dependencies 추가
    for dep, version in DEPENDENCIES.items():
        if _has_dependency(content, dep):
            print(f"  Skip: {dep} (already present)")
            continue
        if dry_run:
            print(f"  [dry-run] Would add {dep}: {version} to dependencies")
            added.append(dep)
        else:
            content = _add_dependency(content, "dependencies", dep, version)
            added.append(dep)

    # dev_dependencies 추가
    for dep, version in DEV_DEPENDENCIES.items():
        if _has_dependency(content, dep):
            print(f"  Skip: {dep} (already present)")
            continue
        if dry_run:
            print(f"  [dry-run] Would add {dep}: {version} to dev_dependencies")
            added.append(dep)
        else:
            content = _add_dependency(content, "dev_dependencies", dep, version)
            added.append(dep)

    if not added:
        print("  All dependencies already present.")
        return False

    if not dry_run and content != original:
        pubspec_path.write_text(content, encoding="utf-8")
        print(f"  Added {len(added)} dependencies: {', '.join(added)}")

    return True


def main() -> int:
    """CLI 진입점."""
    if len(sys.argv) < 2:
        print("Usage: update_pubspec.py <pubspec_path> [--dry-run]", file=sys.stderr)
        return 1

    pubspec_path = Path(sys.argv[1])
    dry_run = "--dry-run" in sys.argv

    update_pubspec(pubspec_path, dry_run)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/flutter/openapi/update-pubspec.mjs
#!/usr/bin/env node
// =============================================================================
// Adds BuiltValue dependencies to a package's pubspec.yaml.
//
// Dependencies:      built_value, built_collection
// Dev dependencies:  build_runner, built_value_generator
//
// Idempotent: skips entries that already exist.
//
// Usage:
//   update-pubspec.mjs <pubspec_path> [--dry-run]
// =============================================================================

import fs from 'node:fs';
import process from 'node:process';

const DEPENDENCIES = {
  built_value: '^8.12.3',
  built_collection: '^5.1.1',
};

const DEV_DEPENDENCIES = {
  build_runner: '^2.4.15',
  built_value_generator: '^8.12.3',
};

const HELP = `Usage: update-pubspec.mjs <pubspec_path> [--dry-run]

Adds BuiltValue dependencies to a package's pubspec.yaml (idempotent).

Arguments:
  <pubspec_path>   Path to pubspec.yaml

Options:
  --dry-run        Preview changes without writing
  -h, --help       Show this help
`;

function usage(code = 1) {
  (code === 0 ? process.stdout : process.stderr).write(HELP);
  process.exit(code);
}

function parseArgs(argv) {
  const rest = argv.slice(2);
  const args = { pubspecPath: '', dryRun: false };

  for (const a of rest) {
    if (a === '-h' || a === '--help') usage(0);
  }

  if (rest.length >= 1 && !rest[0].startsWith('-')) {
    args.pubspecPath = rest.shift();
  }

  while (rest.length > 0) {
    const a = rest.shift();
    switch (a) {
      case '--dry-run':
        args.dryRun = true;
        break;
      default:
        process.stderr.write(`Unknown option: ${a}\n`);
        usage();
    }
  }

  if (!args.pubspecPath) {
    process.stderr.write('Usage: update-pubspec.mjs <pubspec_path> [--dry-run]\n');
    process.exit(1);
  }

  return args;
}

function escapeRegExp(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function findBlockLastEntry(content, blockName) {
  const lines = content.split('\n');
  let inBlock = false;
  let lastEntryIdx = -1;

  const headerRe = new RegExp(`^${escapeRegExp(blockName)}:`);

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (headerRe.test(line)) {
      inBlock = true;
      continue;
    }
    if (inBlock) {
      if (line.startsWith('  ')) {
        lastEntryIdx = i;
      } else if (line.trim() && !line.startsWith(' ')) {
        break;
      }
    }
  }

  return lastEntryIdx;
}

function hasDependency(content, depName) {
  const re = new RegExp(`^\\s+${escapeRegExp(depName)}:`, 'm');
  return re.test(content);
}

function addDependency(content, blockName, depName, version) {
  const lastIdx = findBlockLastEntry(content, blockName);
  if (lastIdx === -1) return content;

  const lines = content.split('\n');
  lines.splice(lastIdx + 1, 0, `  ${depName}: ${version}`);
  return lines.join('\n');
}

function updatePubspec(pubspecPath, dryRun) {
  if (!fs.existsSync(pubspecPath)) {
    process.stderr.write(`Error: ${pubspecPath} not found\n`);
    return false;
  }

  let content = fs.readFileSync(pubspecPath, 'utf8');
  const original = content;
  const added = [];

  for (const [dep, version] of Object.entries(DEPENDENCIES)) {
    if (hasDependency(content, dep)) {
      process.stdout.write(`  Skip: ${dep} (already present)\n`);
      continue;
    }
    if (dryRun) {
      process.stdout.write(`  [dry-run] Would add ${dep}: ${version} to dependencies\n`);
      added.push(dep);
    } else {
      content = addDependency(content, 'dependencies', dep, version);
      added.push(dep);
    }
  }

  for (const [dep, version] of Object.entries(DEV_DEPENDENCIES)) {
    if (hasDependency(content, dep)) {
      process.stdout.write(`  Skip: ${dep} (already present)\n`);
      continue;
    }
    if (dryRun) {
      process.stdout.write(`  [dry-run] Would add ${dep}: ${version} to dev_dependencies\n`);
      added.push(dep);
    } else {
      content = addDependency(content, 'dev_dependencies', dep, version);
      added.push(dep);
    }
  }

  if (added.length === 0) {
    process.stdout.write('  All dependencies already present.\n');
    return false;
  }

  if (!dryRun && content !== original) {
    fs.writeFileSync(pubspecPath, content, 'utf8');
    process.stdout.write(`  Added ${added.length} dependencies: ${added.join(', ')}\n`);
  }

  return true;
}

function main() {
  const args = parseArgs(process.argv);
  updatePubspec(args.pubspecPath, args.dryRun);
  return 0;
}

process.exit(main());
